mina-sshd-api-core/src/main/java/io/jenkins/plugins/mina_sshd_api/core/authenticators/MinaSSHPasswordKeyAuthenticator.java

```diff
@@ -0,0 +1,139 @@
+import com.cloudbees.plugins.credentials.common.StandardUsernamePasswordCredentials;
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import hudson.Extension;
+import hudson.Functions;
+import org.apache.sshd.client.auth.password.UserAuthPasswordFactory;
+import org.apache.sshd.client.session.ClientSession;
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Does password auth with a {@link ClientSession}.
+ */
+public class MinaSSHPasswordKeyAuthenticator 
+    extends SSHAuthenticator<ClientSession, StandardUsernamePasswordCredentials> {
+
+    static /*almost final*/ int authTimeout = Integer.parseInt(
+        System.getProperty(MinaSSHPasswordKeyAuthenticator.class.getName() + ".authTimeout", "15"));
+
+    /**
+     * Constructor.
+     *
+     * @param connection the connection we will be authenticating.
+     * @since 1.4
+     */
+    MinaSSHPasswordKeyAuthenticator(@NonNull ClientSession connection,
+                                    @NonNull StandardUsernamePasswordCredentials user,
+                                    @CheckForNull String username) {
+        super(connection, user, username);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean canAuthenticate() {
+        return getConnection().getUserAuthFactories().stream()
+            .anyMatch(userAuthFactory -> userAuthFactory instanceof UserAuthPasswordFactory)
+            && !getConnection().isAuthenticated()
+            && getConnection().isOpen();
+    }
+
+    @NonNull
+    @Override
+    public Mode getAuthenticationMode() {
+        return Mode.AFTER_CONNECT;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected boolean doAuthenticate() {
+        final String username = getUsername();
+        getConnection().addPasswordIdentity(getUser().getPassword().getPlainText());
+        try {
+            getConnection().setUsername(username);
+            return getConnection().auth().verify(authTimeout, TimeUnit.SECONDS).isSuccess();
+        } catch (IOException e) {
+            Functions.printStackTrace(e, getListener().error("Could not authenticate due to I/O issue"));
+        }
+        return false;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Extension(optional = true)
+    public static class Factory extends SSHAuthenticatorFactory {
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        protected <C, U extends StandardUsernameCredentials> SSHAuthenticator<C, U> newInstance(@NonNull C connection,
+                                                                                                @NonNull U user) {
+            return newInstance(connection, user, null);
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Nullable
+        @Override
+        @SuppressWarnings("unchecked")
```

Comment:
BTW this indicates a design flaw in the `SSHAuthenticator` API. Not sure if it is too late to fix.

Reply:
This is an "unchecked cast" because of the generic (casting `MinaSSHPublicKeyAuthenticator` to `(SSHAuthenticator<C, U>)`). https://github.com/jenkinsci/ssh-credentials-plugin/blob/ssh-credentials-1.19/src/main/java/com/cloudbees/jenkins/plugins/sshcredentials/SSHAuthenticator.java#L226-L246

Reply:
@jglick one way I can think of technically get get rid of the unchecked cast is to maybe generify the [SSHAuthenticatorFactory](https://github.com/jenkinsci/ssh-credentials-plugin/blob/291.v8211e4f8efb_c/src/main/java/com/cloudbees/jenkins/plugins/sshcredentials/SSHAuthenticatorFactory.java) but that would probably introduce breaking API changes.
The [SSHAuthenticatorFactory#support](https://github.com/jenkinsci/ssh-credentials-plugin/blob/291.v8211e4f8efb_c/src/main/java/com/cloudbees/jenkins/plugins/sshcredentials/SSHAuthenticatorFactory.java#L82-L92) method should kind of guarantee that the cast works and therefore justify the unchecked cast ?

Reply:
Not sure offhand. Not particularly relevant here, just something I was wondering about when I saw it.